server, sql: unify sqlInstanceDialer initialization for all tenants

Nukitt opened the pull request:
Previously, system tenants and secondary tenants had different
initialization paths for the ``sqlInstanceDialer``. This was because the 
system tenants used gossip based ``kvNodeDialer`` which uses 
LocalityAddress(--locality-advertise-addr) and defaults to the 
Address (--advertise-addr) of the node (descriptor) and the SQL instance
based address resolver only have the Address field available as they didn't
store locality aware advertised addresses in `system.sql_instances` table.

This could potentially degrade performance in certain cases after the 
unification. Also, we can't use the `kvNodeDialer` as the `sqlInstanceDialer`
for secondary tenants as secondary tenant gRPC servers currently live 
on a different port.

To address this, this patch adds a new column `locality_addresses` to the 
`system.sql_instances` table and then uses it to unify the codepaths for
both the tenants.

Fixes: #138715
Part of: CRDB-46312
Epic: CRDB-46923

Release Note: None

shubhamdhama commented on pkg/sql/sqlinstance/instancestorage/instancestorage.go:
```diff
@@ -313,10 +345,20 @@ func (s *Storage) createInstanceRow(
+				locality,
+				binaryVersion,
+				false,
+				true,
```
I think it should be `encodeLocalityAddress` instead.

Nukitt replied:
correct, fixed it!

shubhamdhama reviewed (changes requested):
Thanks for taking this on. This will be quite useful for both UA efforts and the UA migration.

I've completed the first round of review for this PR's first 2 commits. Additionally, some tests are failing; please fix those and mark the PR as ready for review once resolved.

Nukitt commented on pkg/sql/sqlinstance/instancestorage/instancestorage.go:
```diff
@@ -87,22 +88,23 @@ type Storage struct {
+	locality          roachpb.Locality
+	binaryVersion     roachpb.Version
+	isDraining        bool
+	localityAddresses []roachpb.LocalityAddress
```
done!

Nukitt commented on pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go:
```diff
@@ -90,8 +90,9 @@ func TestGetAvailableInstanceIDForRegion(t *testing.T) {
-				/* isDraining */ false,
+				true,
+				false,
+				nil,
```
done! didnt know about this.

yuzefovich replied:
Yeah, this helps during the review (outside of the IDE) since it's easier to see which arguments simple constants correspond to. We do it on a best effort basis in tests, but in production code it's almost a "must do".
<!-- Reviewable comment -OYXDKJfAoGgiOgaq1ON:b-12xego -->
<!-- Sent from Reviewable.io -->

shubhamdhama reviewed (commented):
So we've found the immediate cause of the `multiregion` test failure. The test drops a region after setting a primary region for the `system` database, and that's causing the `system.sql_instances` table to get wiped out. While it's intuitive, and we are still digging into it and we need to discuss its implications.

```sql
ALTER DATABASE system SET PRIMARY REGION "us-east-1";
ALTER DATABASE d DROP REGION 'us-east-1';
```

Nukitt commented on pkg/upgrade/upgrades/v25_4_sql_instances_add_locality_addresses_test.go:
done 👍

mw5h reviewed (commented):
> @mw5h reviewed 1 of 11 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150730)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb, @Nukitt, and @shubhamdhama)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks! I have some drive-by nits and will defer to Shubham for approval.

> @yuzefovich reviewed 23 of 23 files at r3, 11 of 11 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-:-OYXEQ4x5ioWi_N_XJKK:b50u3zw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb, @Nukitt, and @shubhamdhama)

___
*[`pkg/sql/catalog/systemschema/system.go` line 924 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OYX9m7C5KeGFawzDS5n:-OYX9m7C5KeGFawzDS5o:b-c185ol) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf12e3f5d1ae51d395919f55d58c550bd4ef5fb/pkg/sql/catalog/systemschema/system.go#L924)):*
> ```Go
>     binary_version STRING,
>     is_draining    BOOL NULL,
> 		locality_addresses JSONB NULL,
> ```

nit: the indentation is off here and in the FAMILY clause.
___
*[`pkg/server/server_sql.go` line 321 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OYXDtbF5UhxpFgLe_SW:-OYXDtbF5UhxpFgLe_SX:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1008e7e6b1878296f6ac45b79310b2a571ea21f/pkg/server/server_sql.go#L321)):*
> ```Go
> 	sqlInstanceDialer *nodedialer.Dialer
> 
> 	// Used by sqlInstanceDialer to resolve addresses of SQL instances
> ```

nit: missing period.
___
*[`pkg/server/server_sql.go` line 615 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OYXE6zxCufnD4AeylJw:-OYXE6zxCufnD4AeylJx:b35zslt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1008e7e6b1878296f6ac45b79310b2a571ea21f/pkg/server/server_sql.go#L615)):*
> ```Go
> 	)
> 
> 	// We can't use the nodeDialer as the sqlInstanceDialer unless we
> ```

nit: this comment seems stale now and should probably be removed.
___
*[`pkg/sql/sqlinstance/instancestorage/instancestorage.go` line 234 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OYXBrt79RV-oMh-4Ytn:-OYXBrt79RV-oMh-4Yto:bibrl6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/798a906628122275ecba9bc4a802f16c082827e7/pkg/sql/sqlinstance/instancestorage/instancestorage.go#L234)):*
> ```Go
> 			n.locality,
> 			n.binaryVersion,
> 			true,
> ```

nit: let's keep `true, /* isDraining */` comment.


<!-- Sent from Reviewable.io -->

Nukitt reviewed (commented):
TFTR! 
Addressed the nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-:-OaPdCYl3qaiozqwKTO8:bdaw23m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb and @shubhamdhama)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> @yuzefovich reviewed 22 of 22 files at r6, 11 of 11 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150730)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb, @Nukitt, and @shubhamdhama)

___
*[`pkg/sql/catalog/systemschema/system.go` line 924 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OYX9m7C5KeGFawzDS5n:-OaQLEjG--ABTMuD7HG5:b-j5oxr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf12e3f5d1ae51d395919f55d58c550bd4ef5fb/pkg/sql/catalog/systemschema/system.go#L924)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the indentation is off here and in the FAMILY clause.
</blockquote></details>

nit: the indentation is still off here and in FAMILY clause.
___
*[`pkg/clusterversion/cockroach_versions.go` line 237 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQKtKrDVLyTOelpivL:-OaQKtKs2rWO27OLm40E:b-80ohlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aead4d86f35493cbf7a8bee57966d5517046862/pkg/clusterversion/cockroach_versions.go#L237)):*
> ```Go
> 	// V25_4_SQLInstancesAddLocalityAddresses is the migration to add the `locality_addresses`
> 	// column to the system.sql_instances table.
> 	V25_4_SQLInstancesAddLocalityAddresses
> ```

Given we've already picked beta sha on Friday, we cannot merge this change as is. If we need to get this into 25.4 branch, then we'll need to withdraw beta.1. If we don't need to get it into 25.4, then we'll need to wait some time before we mint 26.1 dev cluster version.
___
*[`pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go` line 93 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQNtPR2qHmG1dllcr2:-OaQNtPR2qHmG1dllcr3:bkhnxzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d7f1aff9bd12c606da685b79267288423150d16/pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go#L93)):*
> ```Go
> 				roachpb.Locality{},
> 				roachpb.Version{},
> 				true,  /* isDraining */
> ```

nit: why did we change `isDraining` from `false` to `true`?
___
*[`pkg/sql/sqlinstance/instancestorage/row_codec.go` line 223 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQMnPf4LjPn7DsQGNi:-OaQMnPf4LjPn7DsQGNj:b7ng8lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d7f1aff9bd12c606da685b79267288423150d16/pkg/sql/sqlinstance/instancestorage/row_codec.go#L223)):*
> ```Go
> ) (*roachpb.Value, error) {
> 	var valueBuf []byte
> 	var encodeErr error
> ```

nit: it'd be slightly easier to read if we changed the signature of `columnsToEncode` functions to have explicit `error` return argument, as opposed to having out-of-scope error.
___
*[`pkg/server/server_sql.go` line 1607 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQOweF0KC09-YeEI7g:-OaQOweG0L2TIoVIaaYj:bk2q0zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/31276ded25017f992959111b76e98247c920244c/pkg/server/server_sql.go#L1607)):*
> ```Go
> 				localityAddresses = nil
> 			}
> 			if err != nil {
> ```

nit: which `err` are we checking here?


<!-- Sent from Reviewable.io -->

Nukitt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150730)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cthumuluru-crdb, @shubhamdhama, and @yuzefovich)

___
*[`pkg/clusterversion/cockroach_versions.go` line 237 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQKtKrDVLyTOelpivL:-OaQnP_h6BhMlpyS-WU_:b-k4wtmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6aead4d86f35493cbf7a8bee57966d5517046862/pkg/clusterversion/cockroach_versions.go#L237)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Given we've already picked beta sha on Friday, we cannot merge this change as is. If we need to get this into 25.4 branch, then we'll need to withdraw beta.1. If we don't need to get it into 25.4, then we'll need to wait some time before we mint 26.1 dev cluster version.
</blockquote></details>

Yeah we're not aiming for 25.4 for this. Will wait for 26.1, thought it would be good to have this ready by then, hence the edits. Maybe I'll move this to drafts till then.
___
*[`pkg/server/server_sql.go` line 1607 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQOweF0KC09-YeEI7g:-OadwghxD6EWgDc7ILbj:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/31276ded25017f992959111b76e98247c920244c/pkg/server/server_sql.go#L1607)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: which `err` are we checking here?
</blockquote></details>

fixed.
___
*[`pkg/sql/catalog/systemschema/system.go` line 924 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OYX9m7C5KeGFawzDS5n:-Oadgd480Iw7NocuSnZ0:b8rch01) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf12e3f5d1ae51d395919f55d58c550bd4ef5fb/pkg/sql/catalog/systemschema/system.go#L924)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the indentation is still off here and in FAMILY clause.
</blockquote></details>

Oh weird. Now should be fixed.
___
*[`pkg/sql/sqlinstance/instancestorage/row_codec.go` line 223 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQMnPf4LjPn7DsQGNi:-OadzuHkEqCtRosURR1i:b-eadllk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d7f1aff9bd12c606da685b79267288423150d16/pkg/sql/sqlinstance/instancestorage/row_codec.go#L223)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it'd be slightly easier to read if we changed the signature of `columnsToEncode` functions to have explicit `error` return argument, as opposed to having out-of-scope error.
</blockquote></details>

Will consider this and take it up for a discussion with the team. Will come back on this.
___
*[`pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go` line 93 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/150730#-OaQNtPR2qHmG1dllcr2:-Oae2CQd1rXLKxJVDQRA:b-qssebl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d7f1aff9bd12c606da685b79267288423150d16/pkg/sql/sqlinstance/instancestorage/instancestorage_internal_test.go#L93)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why did we change `isDraining` from `false` to `true`?
</blockquote></details>

fixed.


<!-- Sent from Reviewable.io -->

Nukitt committed:
sql,upgrade: add `locality_addresses` column to `system.sql_instances` table

Previously, the `system.sql_instances` table didn't store locality
aware advertised addresses.

This change was required to unify the initialization paths of the
`sqlInstanceDialer`, more context on that at #138715.

To address this, this patch adds a new column `locality_addresses`
to the `system.sql_instances` table. This change is also accompanied
by a version gate and a cluster migration to handle the schema upgrade.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312
Release note: None

Nukitt committed:
sql: update codec for `system.sql_instances` table

Previously, the `system.sql_instances` table did not store a list of
locality-aware addresses for SQL instances.

This was inadequate because it created a separate initialization path for
the `sqlInstanceDialer` depending on the tenant type.

To address this, this patch adds the encoding and decoding for the
new `locality_addresses` column. This allows locality-aware addresses
to be persisted directly, which in turn helps unify the dialer's
initialization for all tenants.

Epic: CRDB-46923
Part of: #138715
Part of: CRDB-46312

Release note: None

Nukitt committed:
server: unify `sqlInstanceDialer` initialization

Previously, sqlInstanceDialer was instantiated differently for system
and secondary tenants. For system tenant, kvNodeDialer was reused as
sqlInstanceDialer. For secondary tenants, a custom address resolver
(using sqlInstanceReader) was used to instantiate sqlInstanceDialer.

We needed to unify these initialization paths to reduce the
gap between system and secondary tenants as part of UA migration.
The locality-aware addresses of SQL instances are now stored in the
sql_instances table, which enables unifying the initialization paths.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312
Release note: None
